README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

<p align="center">
   <img src="https://user-images.githubusercontent.com/46109133/148128997-48ee1d39-8488-4227-83d9-6d80f7034a1b.png" alt="adaptable" width="200"/>
</p>

<p align="center">
  <img alt="Languages" src="https://img.shields.io/github/languages/count/Sthaynny/adaptable_screen?color=%235963C5" />
  <img alt="lastcommit" src="https://img.shields.io/github/last-commit/Sthaynny/adaptable_screen?color=%235761C3" />
  <img alt="Issues" src="https://img.shields.io/github/issues/Sthaynny/adaptable_screen?color=%235965E0">

  </a>
</p>

<p>
  
The plug-in scales sizes based on the size of the UIDesign prototypes and the windows the components are being placed into.
  
</p>

## Getting started

Depend on it

Add this to your package's pubspec.yaml file:


```
  dependencies:
       adaptable_screen: ^0.0.4
```


Install it
You can install packages from the command line:

with Flutter:

```
  $ flutter pub get
```

## Usage

To set the widths and heights based on the build size of your design, use the `AdaptableScreenUtils.init` function to set the sizes. Essa função deve ser utilizada ao iniciar a construção do seu aplicativo.

```dart
void main() {
  AdaptableScreenUtils.init(
    heightUIDisign: 1000,
    widthUIDisign: 956,
  );
  runApp(const MyApp());
}
```

Use `.h` to scale the height based on the window size:

```dart
Container(
  height: 10.h
)
```

Use `.w` to scale the width based on the window size:

```dart
Container(
  width: 10.w
)
```

Use `.ssp` to scale text based on window size, this function can change font size as window changes size:

```dart
Text(
  "Text Exemplo",
  style: TextStyle(fontSize: 15.ssp),
)
```

Use `.sp` to scale text based on window size:

```dart
Text(
  "Text Exemplo",
  style: TextStyle(fontSize: 15.ssp),
)
```



Use `.r` to scale radius based on minimum window size:

```dart
Container(
  decoration: BoxDecoration(
    color: Colors.grey,
    borderRadius: BorderRadius.circular(10.r),
  ),
  height: 50.h,
  width: 100.w,
  child: Center(
    child: Text(
      "Text Exemplo",
      style: TextStyle(fontSize: 15.sp),
    ),
  ),
)
```

If you want to see a usage in a simple way, go to `/example` folder.
## Additional information

For contributes:

- Fork this repository;
- Create a new branch to develop your feature: `git checkout -b my-feature`;
- Commit your changes: `git commit -m 'feat: my new feature'`;
- Push to your branch: `git push origin my-feature`.
- Open a pull request for your code to be evaluated.

To help maintain the chosen pattern we also create a file which is called before every commit. This file will format and pinpoint (if present) errors in the codestyle of your code. To enable this you must first copy it to git's hooks folder. If you are developing on macOS, go to the root of the project and run the command below:

```
cp pre-commit .git/hooks/pre-commit
```

After this step, it is necessary to give permission for the file to be executed. Just follow the following command:

```
chmod +x .git/hooks/pre-commit
```


CHANGELOG.md
## 0.0.1

In this first version was created:
- Resize screens based on an interface size;
- Resize the embroidery with .r;
- Creation of function to initialize the base width and height to be resized;
- Adapting for Null Safety

## 0.0.2

Update description.

## 0.0.3

Function update to scale the radius of the .r function 

## 0.0.4

- Update documentation.
- Function call optimization to facilitate imports
- Update package usage description

lib/src/adaptable_screen_utils.dart
part of adaptable_screen;

class AdaptableScreenUtils {
  static double? _heightUIDisign;
  static double? _widthUIDisign;

  static final AdaptableScreenUtils _instance =
      AdaptableScreenUtils._internal();

  factory AdaptableScreenUtils() {
    return _instance;
  }

  AdaptableScreenUtils._internal();

  static void init({double? heightUIDisign, double? widthUIDisign}) {
    assert(heightUIDisign != 0);
    assert(widthUIDisign != 0);
    _heightUIDisign = heightUIDisign;
    _widthUIDisign = widthUIDisign;
  }

  /// Size of the phone in UI Design , px

  Size get _defaultSize => Size(_widthUIDisign ?? 375, _heightUIDisign ?? 812);

  /// Representing the main window for applications where there is only one
  /// window, such as applications designed for single-display mobile windows or
  ///  the main window to which this binding is bound.
  SingletonFlutterWindow get _window =>
      WidgetsBinding.instance?.window ?? ui.window;

  /// Width of the main window that this link is linked to.
  double get width {
    return MediaQueryData.fromWindow(_window).size.width;
  }

  /// Height of the main window that this link is linked to.
  double get height {
    return MediaQueryData.fromWindow(_window).size.height;
  }

  /// Represents the actual window scaling factor.
  double get factorTextScaleRealwindow {
    return MediaQueryData.fromWindow(_window).textScaleFactor;
  }

  // Represents the scale factor with the range up to 1.6
  double get factorTextScale {
    return factorTextScaleRealwindow <= 1.6 ? factorTextScaleRealwindow : 1.6;
  }

  /// The ratio of the actual dp to the design draft px
  double get scaleWidth => width / _defaultSize.width;

  /// The ratio of the actual dp to the design draft px
  double get scaleHeiht => height / _defaultSize.height;

  /// text scaled based on width.
  double get textScale => scaleWidth;

  /// The offset from the top
  double get heightStatusBar => MediaQueryData.fromWindow(_window).padding.top;

  /// The offset from the bottom.
  double get heightBottomBar =>
      MediaQueryData.fromWindow(_window).padding.bottom;

  /// Maximum window Height and Width Sizes
  double get lengthMax {
    return max(height, width);
  }

  /// Minimum window Height and Width Sizes
  double get lengthMin {
    return min(height, width);
  }

  /// Scales radius based on minimum window sizes
  double radius(double radius) => radius * atan2(height, width);

  /// Scaled height based on window size.
  double steppedHeight(double height) {
    return height * scaleHeiht;
  }

  /// Scaled width based on window size.
  double steppedWidth(double width) {
    return width * scaleWidth;
  }

  /// Adjusting font according to the scaling factor, the scaling can be dynamic or fixed.
  double setFont(double fontSize, {bool autoScale = false}) {
    if (autoScale) {
      return fontSize * textScale * factorTextScale / factorTextScaleRealwindow;
    }
    return (fontSize * textScale) / factorTextScaleRealwindow;
  }
}


example/lib/main.dart
import 'package:adaptable_screen/adaptable_screen.dart';
import 'package:flutter/material.dart';

void main() {
  AdaptableScreenUtils.init(
    heightUIDisign: 1000,
    widthUIDisign: 956,
  );
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          children: [
            Container(
              decoration: BoxDecoration(
                color: Colors.red,
                borderRadius: BorderRadius.circular(33.r),
              ),
              height: 160.h,
              width: 200.w,
              child: Center(
                child: Text(
                  "Text Exemplo",
                  style: TextStyle(fontSize: 20.ssp),
                ),
              ),
            ),
            Container(
              decoration: BoxDecoration(
                color: Colors.grey,
                borderRadius: BorderRadius.circular(10.r),
              ),
              height: 50.h,
              width: 100.w,
              child: Center(
                child: Text(
                  "Text Exemplo",
                  style: TextStyle(fontSize: 15.sp),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
